File: src/client/java/com/cxxcxx/zinecraft/core/client/weapon/WeaponPresentationController.java
package com.cxxcxx.zinecraft.core.client.weapon;

import com.cxxcxx.zinecraft.api.weapon.WeaponPresentation;
import com.cxxcxx.zinecraft.api.weapon.network.WeaponActionCancelledPayload;
import com.cxxcxx.zinecraft.api.weapon.network.WeaponActionStartedPayload;
import com.cxxcxx.zinecraft.api.weapon.network.WeaponPayloadTypes;
import com.cxxcxx.zinecraft.core.Zinecraft;
import net.minecraft.client.Minecraft;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.item.ItemStack;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.client.event.ClientPlayerNetworkEvent;
import net.neoforged.neoforge.client.event.ClientTickEvent;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

@EventBusSubscriber(modid = Zinecraft.MOD_ID, value = Dist.CLIENT)
public final class WeaponPresentationController {
  private static final Map<Key, ActivePresentation> ACTIVE = new HashMap<>();
  private static final WeaponPresentationVfxService VFX = WeaponVfxServices.create();
  private static final WeaponSoundService SOUNDS = VanillaWeaponSoundService.INSTANCE;
  private static final WeaponAnimationService WEAPON_ANIMATIONS = NoopWeaponAnimationService.INSTANCE;
  private static final PlayerAnimationService PLAYER_ANIMATIONS = VanillaPlayerAnimationService.INSTANCE;

  private WeaponPresentationController() {
  }

  public static void initialize() {
    WeaponPayloadTypes.setClientHandlers(WeaponPresentationController::start, WeaponPresentationController::cancel);
  }

  private static void start(WeaponActionStartedPayload payload) {
    var definition = Zinecraft.WEAPONS.definition(payload.getWeaponId());
    if (definition == null) return;
    var timeline = definition.presentation(payload.getActionId());
    if (timeline != null) ACTIVE.put(new Key(payload.getEntityId(), payload.getActionId()),
        new ActivePresentation(payload.getStartGameTick(), timeline));
  }

  private static void cancel(WeaponActionCancelledPayload payload) {
    var running = ACTIVE.remove(new Key(payload.getEntityId(), payload.getActionId()));
    if (running != null) stop(payload.getEntityId(), running);
  }

  @SubscribeEvent
  public static void tick(ClientTickEvent.Post event) {
    Minecraft client = Minecraft.getInstance();
    var level = client.level;
    if (level == null) return;
    Iterator<Map.Entry<Key, ActivePresentation>> iterator = ACTIVE.entrySet().iterator();
    while (iterator.hasNext()) {
      var entry = iterator.next();
      var running = entry.getValue();
      var entity = level.getEntity(entry.getKey().entityId()) instanceof LivingEntity living ? living : null;
      long elapsed = level.getGameTime() - running.startGameTick;
      if (elapsed >= running.timeline.getDurationTicks()) {
        if (entity != null) stop(entity, running);
        iterator.remove();
        continue;
      }
      if (entity == null || elapsed < 0) continue;
      if (!running.started) {
        ItemStack stack = entity.getMainHandItem();
        if (running.timeline.getWeaponAnimation() != null)
          WEAPON_ANIMATIONS.play(entity, stack, running.timeline.getWeaponAnimation());
        if (running.timeline.getPlayerAnimation() != null)
          PLAYER_ANIMATIONS.play(entity, running.timeline.getPlayerAnimation());
        running.stack = stack;
        running.started = true;
      }
      for (int i = 0; i < running.timeline.getVfx().size(); i++) {
        var cue = running.timeline.getVfx().get(i);
        if (!running.playedVfx[i] && elapsed >= cue.getTick()) {
          VFX.play(entity, cue.getId());
          running.playedVfx[i] = true;
        }
      }
      for (int i = 0; i < running.timeline.getSounds().size(); i++) {
        var cue = running.timeline.getSounds().get(i);
        if (!running.playedSounds[i] && elapsed >= cue.getTick()) {
          SOUNDS.play(entity, cue.getId());
          running.playedSounds[i] = true;
        }
      }
    }
  }

  @SubscribeEvent
  public static void logout(ClientPlayerNetworkEvent.LoggingOut event) {
    ACTIVE.clear();
  }

  private static void stop(int entityId, ActivePresentation running) {
    var level = Minecraft.getInstance().level;
    if (level != null && level.getEntity(entityId) instanceof LivingEntity living) stop(living, running);
  }

  private static void stop(LivingEntity entity, ActivePresentation running) {
    if (!running.started) return;
    ItemStack stack = running.stack == null ? entity.getMainHandItem() : running.stack;
    if (running.timeline.getWeaponAnimation() != null)
      WEAPON_ANIMATIONS.stop(entity, stack, running.timeline.getWeaponAnimation());
    if (running.timeline.getPlayerAnimation() != null)
      PLAYER_ANIMATIONS.stop(entity, running.timeline.getPlayerAnimation());
  }

  private record Key(int entityId, ResourceLocation actionId) {
  }

  private static final class ActivePresentation {
    private final long startGameTick;
    private final WeaponPresentation timeline;
    private final boolean[] playedVfx;
    private final boolean[] playedSounds;
    private boolean started;
    private ItemStack stack;

    private ActivePresentation(long startGameTick, WeaponPresentation timeline) {
      this.startGameTick = startGameTick;
      this.timeline = timeline;
      this.playedVfx = new boolean[timeline.getVfx().size()];
      this.playedSounds = new boolean[timeline.getSounds().size()];
    }
  }
}

File: src/client/kotlin/com/cxxcxx/zinecraft/core/client/weapon/WeaponPresentationController.kt
package com.cxxcxx.zinecraft.core.client.weapon

import com.cxxcxx.zinecraft.api.weapon.WeaponPresentation
import com.cxxcxx.zinecraft.api.weapon.network.WeaponActionCancelledPayload
import com.cxxcxx.zinecraft.api.weapon.network.WeaponActionStartedPayload
import com.cxxcxx.zinecraft.core.Zinecraft
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.minecraft.client.Minecraft
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.LivingEntity
import com.cxxcxx.zinecraft.core.client.weapon.tacz.TaczWeaponAnimationService
import com.cxxcxx.zinecraft.core.client.weapon.tacz.TaczPlayerAnimationService

object WeaponPresentationController {
  private data class Key(val entityId: Int, val actionId: ResourceLocation)

  private data class ActivePresentation(
    val startGameTick: Long,
    val timeline: WeaponPresentation,
    var animationStarted: Boolean = false,
    val playedVfx: BooleanArray = BooleanArray(timeline.vfx.size),
    val playedSounds: BooleanArray = BooleanArray(timeline.sounds.size)
  )

  private val active = mutableMapOf<Key, ActivePresentation>()
  private val playerAnimations: PlayerAnimationService = TaczPlayerAnimationService
  private val weaponAnimations: WeaponAnimationService = TaczWeaponAnimationService
  private val vfx: WeaponVfxService = VanillaWeaponVfxService
  private val sounds: WeaponSoundService = VanillaWeaponSoundService

  fun initialize() {
    ClientPlayNetworking.registerGlobalReceiver(WeaponActionStartedPayload.TYPE) { payload, context ->
      context.client().execute { start(payload) }
    }
    ClientPlayNetworking.registerGlobalReceiver(WeaponActionCancelledPayload.TYPE) { payload, context ->
      context.client().execute { cancel(payload) }
    }
    ClientTickEvents.END_CLIENT_TICK.register(::tick)
    ClientPlayConnectionEvents.DISCONNECT.register { _, _ -> active.clear() }
  }

  private fun start(payload: WeaponActionStartedPayload) {
    val definition = Zinecraft.WEAPONS.definition(payload.weaponId) ?: return
    val timeline = definition.presentation(payload.actionId) ?: return
    active[Key(payload.entityId, payload.actionId)] = ActivePresentation(
      payload.startGameTick,
      timeline
    )
  }

  private fun cancel(payload: WeaponActionCancelledPayload) {
    active.remove(Key(payload.entityId, payload.actionId))?.let { stop(payload.entityId, it) }
  }

  private fun tick(client: Minecraft) {
    val level = client.level ?: return
    val iterator = active.iterator()
    while (iterator.hasNext()) {
      val (key, running) = iterator.next()
      val entity = level.getEntity(key.entityId) as? LivingEntity
      val elapsed = level.gameTime - running.startGameTick
      if (elapsed >= running.timeline.durationTicks) {
        if (entity != null) stop(entity, running)
        iterator.remove()
        continue
      }
      if (entity == null || elapsed < 0) continue

      if (!running.animationStarted) {
        running.timeline.playerAnimation?.let { playerAnimations.play(entity, it) }
        running.timeline.weaponAnimation?.let { weaponAnimations.play(entity, entity.mainHandItem, it) }
        running.animationStarted = true
      }
      running.timeline.vfx.forEachIndexed { index, cue ->
        if (!running.playedVfx[index] && elapsed >= cue.tick) {
          vfx.play(entity, cue.id)
          running.playedVfx[index] = true
        }
      }
      running.timeline.sounds.forEachIndexed { index, cue ->
        if (!running.playedSounds[index] && elapsed >= cue.tick) {
          sounds.play(entity, cue.id)
          running.playedSounds[index] = true
        }
      }
    }
  }

  private fun stop(entityId: Int, running: ActivePresentation) {
    val entity = Minecraft.getInstance().level?.getEntity(entityId) as? LivingEntity ?: return
    stop(entity, running)
  }

  private fun stop(entity: LivingEntity, running: ActivePresentation) {
    running.timeline.playerAnimation?.let { playerAnimations.stop(entity, it) }
    running.timeline.weaponAnimation?.let { weaponAnimations.stop(entity, entity.mainHandItem, it) }
  }
}
